dotnet-authserver/src/TeacherIdentity.AuthServer/Journeys/TrnTokenHelper.cs

```diff
@@ -0,0 +1,171 @@
+{
+    private readonly TeacherIdentityServerDbContext _dbContext;
+    private readonly IClock _clock;
+    private readonly IDqtApiClient _dqtApiClient;
+    private readonly IConfiguration _configuration;
+    private readonly IUserSearchService _userSearchService;
+
+    public TrnTokenHelper(
+        TeacherIdentityServerDbContext dbContext,
+        IClock clock,
+        IDqtApiClient dqtApiClient,
+        IConfiguration configuration,
+        IUserSearchService userSearchService)
+    {
+        _dbContext = dbContext;
+        _clock = clock;
+        _dqtApiClient = dqtApiClient;
+        _configuration = configuration;
+        _userSearchService = userSearchService;
+    }
+
+    public async Task<EnhancedTrnToken?> GetValidTrnToken(OpenIddictRequest request)
+    {
+        var requestedTrnToken = request["trn_token"];
+
+        if (_configuration.GetValue("RegisterWithTrnTokenEnabled", false) && requestedTrnToken.HasValue)
+        {
+            return await GetValidTrnTokenModel(requestedTrnToken.Value.Value as string);
+        }
+
+        return null;
+    }
+
+    public async Task InitializeAuthenticationStateWithToken(
+        User? signedInUser,
+        AuthenticationState authenticationState,
+        EnhancedTrnToken trnToken,
+        HttpContext httpContext)
+    {
+        if (signedInUser is not null)
+        {
+            if (signedInUser.EmailAddress == trnToken.Email)
+            {
+                authenticationState.OnSignedInUserProvided(signedInUser);
+                if (signedInUser.Trn is null)
+                {
+                    authenticationState.OnTrnTokenProvided(trnToken);
+                }
+            }
+            else
+            {
+                authenticationState.OnTrnTokenProvided(trnToken);
+            }
+        }
+        else
+        {
+            var existingValidUser = await _dbContext.Users.SingleOrDefaultAsync(u => u.EmailAddress == trnToken.Email);
+
+            if (existingValidUser is not null)
+            {
+                if (existingValidUser.Trn is not null && existingValidUser.Trn != trnToken.Trn)
+                {
+                    // If we find a matching email in our records but TRN no match, ignore the token
+                    return;
+                }
+
+                authenticationState.OnSignedInUserProvided(existingValidUser);
+                if (existingValidUser.Trn is null)
+                {
+                    authenticationState.OnTrnTokenProvided(trnToken);
+                }
+                await authenticationState.SignIn(httpContext);
+            }
+            else
+            {
+                var users = await _userSearchService.FindUsers(trnToken.FirstName, trnToken.LastName, trnToken.DateOfBirth);
+                authenticationState.OnExistingAccountSearch(users.Length == 0 ? null : users[0]);
+                authenticationState.OnTrnTokenProvided(trnToken);
+            }
+        }
+    }
```

Comment:
I think we can sacrifice readability to reduce nesting here, but not sure

Reply:
I think it's fine as-is.